### ISSUE TITLE
[next][RC] Allow plugins to work in Workers too

### ISSUE THREAD

**Issue by WebReflection** [pyscript/pyscript#1678]
As the *worker* story is fairly new, and all PyScript plugins to date assumed to work only on the main thread, the current logic works well for main thread case but it needs some orchestration for the workers, as worker code can't be a file to import lazily and don't bootstrap the same way the main page does, simply because they live and operate in a different "world" that is not the main thread.

This issue is here to both remind everyone that:

  * workers plugins can't work out of the box as it is for main because they can't be loaded remotely (workers are a Blob with no origin for security issues and implications around that pattern)
  * workers DOM is not globally available like it is given for granted to main thread plugins (they require artifacts or bundler orchestration to assume `document` or even `window` exist)
  * anything needs to somehow propagate in the main thread, *without* requiring the main thread to bootstrap an interpreter, or the main use to desire workers (not having to block the main thread for our interpreter) would be erased

### Additional Context

I don't think this can ever make it for this week release and it requires some thinking and proper logic/orchestration to make it in the near future but at the same time I think it's possible, although I am worried all core plugins might inevitably land within core as embedded code for workers or, even worse, foreign plugins won't be easily achievable due headers constraints for the worker story ... we need to think this through.

**Comment by tedpatrick** [comment]
Do we need plugins for workers? I can see the use case for plugins in main (components/connected apps) but in workers plugins likely are best as modules. As long as we can `import` modules in a worker, I feel that is good enough.

**Comment by ntoll** [comment]
I think we should wait and see... as in, if users start to ask for worker support for plugins, we know we should handle them. Otherwise, we've dodged a bullet.

TL;DR - let's see what folks do with workers and refine from actual evidence / feedback.

**Comment by WebReflection** [comment]
to both: I don't think we need anything for our first release and I agree we should see how, if, and when, workers will be used more to listen to our users and see what, if, and when, we can do something about it.

This is here just as reminder and as link to point at, it's not a blocker, but it's definitively a must know.

**Comment by WebReflection** [comment]
@fpliger 

> I also think that there will be some (internal?) plugins that are expected to work on workers (the error banner on DOM is a good example for instance)

this is why I've opened this issue ... as I can see that coming and I want everyone to understand plugins on main can't automagically translate into plugins on workers ... once we have a single exception it means:

  * that's either not a plugin anymore, it's something else due us not having a solution
  * if it's a plugin and it can work cross realms / environments, it means all plugins can too

This is why we need to be careful here in thinking, or even promoting, the `worker` story as if it's just the same but non-blocking ... that's probably the case for various existent use cases, but for "old style plugin" that can't be the case at all and we need to think about it.

- - -

Side note: the previous *stdlib* (or better, the *display.py*) was after orchestration and we didn't like it already ... that's why I think we need to be sure we do the right thing here but we *could* use an approach similar to the one used in the `stdlib` already and see if that works.

That approach requires every plugin to use always the same entry point to retrieve `document`, `window`, `IS_WORKER` or other things, so that we can disambiguate as much as we like, still the embedding plugins code in the Worker blob somehow bothers me as imports can't easily work there out of the box.

Happy to discuss this elsewhere and properly or even experiment solutions but if we want to move forward the release I won't have time to do so.

**Comment by fpliger** [comment]
> This is why we need to be careful here in thinking, or even promoting, the worker story as if it's just the same but non-blocking ... that's probably the case for various existent use cases, but for "old style plugin" that can't be the case at all and we need to think about it.

Can you elaborate a bit on what you mean? Not sure I fully understand...

> That approach requires every plugin to use always the same entry point to retrieve document, window, IS_WORKER or other things, so that we can disambiguate as much as we like

Honestly, think this should just be a requirement for plugins to work correctly

**Comment by WebReflection** [comment]
I think latest plugins already show that is possible to create plugins that work on both main and worker via dedicated hooks.

I am closing this as done. We might want to open a new issue based on current gotchas, if any.
